File: mobile/sharedLogic/src/commonTest/kotlin/org/example/project/FoodsClientTest.kt
package org.example.project

import io.ktor.client.HttpClient
import io.ktor.client.engine.mock.MockEngine
import io.ktor.client.engine.mock.respond
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpMethod
import io.ktor.http.HttpStatusCode
import io.ktor.http.headersOf
import io.ktor.serialization.kotlinx.json.json
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.Json

class FoodsClientTest {

    @Test
    fun listSendsBearerAndParsesFoods() =
        runTest {
            val store = InMemoryTokenStore()
            store.saveToken("tok", rememberMe = true)
            var sawAuth: String? = null
            val engine =
                MockEngine { request ->
                    assertEquals(HttpMethod.Get, request.method)
                    assertEquals("/api/foods", request.url.encodedPath)
                    sawAuth = request.headers[HttpHeaders.Authorization]
                    respond(
                        content =
                            """
                            [{"id":"cccccccc-cccc-cccc-cccc-cccccccccccc",
                              "name":"Extra apple mash",
                              "iconKey":"apple",
                              "householdId":"22222222-2222-2222-2222-222222222222",
                              "system":false,
                              "sessionEligible":true,
                              "archivedAt":null}]
                            """.trimIndent(),
                        status = HttpStatusCode.OK,
                        headers = headersOf(HttpHeaders.ContentType, "application/json"),
                    )
                }

            val client = FoodsClient("http://localhost:8080", httpClient(engine), store)
            val foods = client.list()

            assertEquals(1, foods.size)
            assertEquals("Extra apple mash", foods[0].name)
            assertFalse(foods[0].system)
            assertEquals(true, foods[0].sessionEligible)
            assertNull(foods[0].iconUrl)
            assertEquals("Bearer tok", sawAuth)
        }

    @Test
    fun listParsesIconUrlWhenPresent() =
        runTest {
            val store = InMemoryTokenStore()
            store.saveToken("tok", rememberMe = true)
            val engine =
                MockEngine { request ->
                    assertEquals(HttpMethod.Get, request.method)
                    respond(
                        content =
                            """
                            [{"id":"cccccccc-cccc-cccc-cccc-cccccccccccc",
                              "name":"Cucumber",
                              "iconKey":"custom_cucumber",
                              "iconUrl":"https://cdn.example.com/foods/custom_cucumber.png",
                              "householdId":"22222222-2222-2222-2222-222222222222",
                              "system":false,
                              "sessionEligible":true,
                              "archivedAt":null}]
                            """.trimIndent(),
                        status = HttpStatusCode.OK,
                        headers = headersOf(HttpHeaders.ContentType, "application/json"),
                    )
                }

            val client = FoodsClient("http://localhost:8080", httpClient(engine), store)
            val foods = client.list()

            assertEquals(
                "https://cdn.example.com/foods/custom_cucumber.png",
                foods[0].iconUrl,
            )
        }

    @Test
    fun createSnackSendsPreferences() =
        runTest {
            val store = InMemoryTokenStore()
            store.saveToken("tok", rememberMe = true)
            val engine =
                MockEngine { request ->
                    assertEquals(HttpMethod.Post, request.method)
                    assertEquals("/api/foods", request.url.encodedPath)
                    respond(
                        content =
                            """
                            {"id":"cccccccc-cccc-cccc-cccc-cccccccccccc",
                             "name":"Salt chips",
                             "iconKey":"custom_chips",
                             "householdId":"22222222-2222-2222-2222-222222222222",
                             "system":false,
                             "sessionEligible":false,
                             "liked":"like",
                             "texture":"crunchy",
                             "tasteNote":"salt & vinegar",
                             "archivedAt":null}
                            """.trimIndent(),
                        status = HttpStatusCode.Created,
                        headers = headersOf(HttpHeaders.ContentType, "application/json"),
                    )
                }

            val client = FoodsClient("http://localhost:8080", httpClient(engine), store)
            val created =
                client.create(
                    name = "Salt chips",
                    iconKey = "custom_chips",
                    sessionEligible = false,
                    liked = "like",
                    texture = "crunchy",
                    tasteNote = "salt & vinegar",
                )

            assertEquals(false, created.sessionEligible)
            assertEquals("like", created.liked)
            assertEquals("crunchy", created.texture)
            assertEquals("salt & vinegar", created.tasteNote)
        }

    @Test
    fun upsertAndClearExposureHitExpectedPaths() =
        runTest {
            val store = InMemoryTokenStore()
            store.saveToken("tok", rememberMe = true)
            val paths = mutableListOf<String>()
            val engine =
                MockEngine { request ->
                    paths += "${request.method.value} ${request.url.encodedPath}?${request.url.encodedQuery}"
                    if (request.method == HttpMethod.Delete) {
                        respond("", status = HttpStatusCode.NoContent)
                    } else {
                        respond(
                            content =
                                """
                                {"foodId":"cccccccc-cccc-cccc-cccc-cccccccccccc",
                                 "variantKey":"bagelsaurus",
                                 "familiarity":"safe",
                                 "source":"manual"}
                                """.trimIndent(),
                            status = HttpStatusCode.OK,
                            headers = headersOf(HttpHeaders.ContentType, "application/json"),
                        )
                    }
                }

            val client = FoodsClient("http://localhost:8080", httpClient(engine), store)
            val upserted =
                client.upsertExposure(
                    foodId = "cccccccc-cccc-cccc-cccc-cccccccccccc",
                    familiarity = "safe",
                    variantKey = "Bagelsaurus",
                )
            client.clearExposure("cccccccc-cccc-cccc-cccc-cccccccccccc", "Bagelsaurus")

            assertEquals("bagelsaurus", upserted.variantKey)
            assertEquals("safe", upserted.familiarity)
            assertEquals(
                listOf(
                    "PUT /api/foods/cccccccc-cccc-cccc-cccc-cccccccccccc/exposures?",
                    "DELETE /api/foods/cccccccc-cccc-cccc-cccc-cccccccccccc/exposures?variantKey=Bagelsaurus",
                ),
                paths,
            )
        }

    @Test
    fun bootstrapSafesHitsExpectedPathAndParsesSignupSource() =
        runTest {
            val store = InMemoryTokenStore()
            store.saveToken("tok", rememberMe = true)
            var path = ""
            val engine =
                MockEngine { request ->
                    path = "${request.method.value} ${request.url.encodedPath}"
                    respond(
                        content =
                            """
                            [{"foodId":"aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaa04",
                              "variantKey":"honeycrisp",
                              "familiarity":"safe",
                              "source":"signup"}]
                            """.trimIndent(),
                        status = HttpStatusCode.OK,
                        headers = headersOf(HttpHeaders.ContentType, "application/json"),
                    )
                }

            val client = FoodsClient("http://localhost:8080", httpClient(engine), store)
            val results =
                client.bootstrapSafes(
                    listOf(
                        BootstrapSafeItemRequest(
                            name = "Apples",
                            variantKey = "Honeycrisp",
                            sessionEligible = true,
                        ),
                        BootstrapSafeItemRequest(name = "Goldfish", sessionEligible = false),
                    ),
                )

            assertEquals("POST /api/foods/bootstrap-safes", path)
            assertEquals(1, results.size)
            assertEquals("signup", results[0].source)
            assertEquals("honeycrisp", results[0].variantKey)
            assertEquals("aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaa04", results[0].foodId)
        }

    @Test
    fun createUpdateArchiveHitExpectedPaths() =
        runTest {
            val store = InMemoryTokenStore()
            store.saveToken("tok", rememberMe = true)
            val paths = mutableListOf<String>()
            val engine =
                MockEngine { request ->
                    paths += "${request.method.value} ${request.url.encodedPath}"
                    val body =
                        """
                        {"id":"cccccccc-cccc-cccc-cccc-cccccccccccc",
                         "name":"Food",
                         "iconKey":"apple",
                         "householdId":"22222222-2222-2222-2222-222222222222",
                         "system":false,
                         "sessionEligible":true,
                         "archivedAt":null}
                        """.trimIndent()
                    val status =
                        if (request.method == HttpMethod.Post &&
                            request.url.encodedPath == "/api/foods"
                        ) {
                            HttpStatusCode.Created
                        } else {
                            HttpStatusCode.OK
                        }
                    respond(
                        content = body,
                        status = status,
                        headers = headersOf(HttpHeaders.ContentType, "application/json"),
                    )
                }

            val client = FoodsClient("http://localhost:8080", httpClient(engine), store)
            client.create("Food", "apple")
            client.update("cccccccc-cccc-cccc-cccc-cccccccccccc", name = "Renamed")
            client.archive("cccccccc-cccc-cccc-cccc-cccccccccccc")

            assertEquals(
                listOf(
                    "POST /api/foods",
                    "PUT /api/foods/cccccccc-cccc-cccc-cccc-cccccccccccc",
                    "POST /api/foods/cccccccc-cccc-cccc-cccc-cccccccccccc/archive",
                ),
                paths,
            )
        }

    @Test
    fun invalidIconSurfacesApiMessage() =
        runTest {
            val store = InMemoryTokenStore()
            store.saveToken("tok", rememberMe = true)
            val engine =
                MockEngine {
                    respond(
                        content = """{"message":"Invalid icon key"}""",
                        status = HttpStatusCode.BadRequest,
                        headers = headersOf(HttpHeaders.ContentType, "application/json"),
                    )
                }

            val client = FoodsClient("http://localhost:8080", httpClient(engine), store)
            val error =
                assertFailsWith<FoodsException> {
                    client.create("Bad", "nope")
                }
            assertEquals("Invalid icon key", error.message)
        }

    @Test
    fun duplicateNameSurfacesApiMessage() =
        runTest {
            val store = InMemoryTokenStore()
            store.saveToken("tok", rememberMe = true)
            val engine =
                MockEngine {
                    respond(
                        content = """{"message":"A food with that name already exists"}""",
                        status = HttpStatusCode.Conflict,
                        headers = headersOf(HttpHeaders.ContentType, "application/json"),
                    )
                }

            val client = FoodsClient("http://localhost:8080", httpClient(engine), store)
            val error =
                assertFailsWith<FoodsException> {
                    client.create("Watermelon", "custom_watermelon")
                }
            assertEquals("A food with that name already exists", error.message)
        }

    @Test
    fun requiresSignedInToken() =
        runTest {
            val client =
                FoodsClient(
                    "http://localhost:8080",
                    httpClient(MockEngine { error("unreachable") }),
                    InMemoryTokenStore(),
                )
            val error = assertFailsWith<FoodsException> { client.list() }
            assertEquals("Not signed in", error.message)
        }

    private fun httpClient(engine: MockEngine): HttpClient =
        HttpClient(engine) {
            install(ContentNegotiation) {
                json(Json { ignoreUnknownKeys = true })
            }
        }
}

File: web/src/api/foodsClient.test.ts
import { describe, expect, it, vi } from "vitest"

import { FoodsClient } from "@/api/foodsClient"
import type { TokenStore } from "@/api/tokenStore"

function memoryStore(token: string | null = "tok"): TokenStore {
  let stored = token
  return {
    get: () => stored,
    set: (value) => {
      stored = value
    },
    clear: () => {
      stored = null
    },
  }
}

const sampleFood = {
  id: "cccccccc-cccc-cccc-cccc-cccccccccccc",
  name: "Extra apple mash",
  iconKey: "apple",
  householdId: "22222222-2222-2222-2222-222222222222",
  system: false,
  sessionEligible: true,
  liked: null,
  texture: null,
  tasteNote: null,
  archivedAt: null,
  exposures: [],
}

describe("FoodsClient", () => {
  it("lists foods with bearer auth", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(JSON.stringify([sampleFood]), {
        status: 200,
        headers: { "Content-Type": "application/json" },
      }),
    )

    const client = new FoodsClient("http://localhost:8080", fetchFn, memoryStore())
    const foods = await client.list()

    expect(foods).toHaveLength(1)
    expect(String(fetchFn.mock.calls[0]?.[0])).toBe(
      "http://localhost:8080/api/foods",
    )
    const init = fetchFn.mock.calls[0]?.[1] as RequestInit
    expect(new Headers(init.headers).get("Authorization")).toBe("Bearer tok")
  })

  it("lists with includeArchived query", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(JSON.stringify([]), {
        status: 200,
        headers: { "Content-Type": "application/json" },
      }),
    )

    const client = new FoodsClient("http://localhost:8080", fetchFn, memoryStore())
    await client.list(true)

    expect(String(fetchFn.mock.calls[0]?.[0])).toBe(
      "http://localhost:8080/api/foods?includeArchived=true",
    )
  })

  it("creates a household food", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(JSON.stringify(sampleFood), {
        status: 201,
        headers: { "Content-Type": "application/json" },
      }),
    )

    const client = new FoodsClient("http://localhost:8080", fetchFn, memoryStore())
    const created = await client.create({ name: "Extra apple mash", iconKey: "apple" })

    expect(created.name).toBe("Extra apple mash")
    const init = fetchFn.mock.calls[0]?.[1] as RequestInit
    expect(init.method).toBe("POST")
    expect(init.body).toBe(
      JSON.stringify({ name: "Extra apple mash", iconKey: "apple" }),
    )
  })

  it("creates a snack with preferences", async () => {
    const snack = {
      ...sampleFood,
      name: "Salt chips",
      sessionEligible: false,
      liked: "like" as const,
      texture: "crunchy" as const,
      tasteNote: "salt & vinegar",
    }
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(JSON.stringify(snack), {
        status: 201,
        headers: { "Content-Type": "application/json" },
      }),
    )
    const client = new FoodsClient("http://localhost:8080", fetchFn, memoryStore())
    const request = {
      name: "Salt chips",
      iconKey: "custom_chips",
      sessionEligible: false,
      liked: "like" as const,
      texture: "crunchy" as const,
      tasteNote: "salt & vinegar",
    }

    const created = await client.create(request)

    expect(created.sessionEligible).toBe(false)
    expect(created.tasteNote).toBe("salt & vinegar")
    expect((fetchFn.mock.calls[0]?.[1] as RequestInit).body).toBe(
      JSON.stringify(request),
    )
  })

  it("updates and archives by id", async () => {
    const fetchFn = vi
      .fn()
      .mockResolvedValueOnce(
        new Response(JSON.stringify({ ...sampleFood, name: "Updated" }), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
      )
      .mockResolvedValueOnce(
        new Response(
          JSON.stringify({ ...sampleFood, archivedAt: "2026-07-14T00:00:00Z" }),
          { status: 200, headers: { "Content-Type": "application/json" } },
        ),
      )

    const client = new FoodsClient("http://localhost:8080", fetchFn, memoryStore())
    await client.update(sampleFood.id, { name: "Updated", iconKey: "sweet_potato" })
    await client.archive(sampleFood.id)

    expect(String(fetchFn.mock.calls[0]?.[0])).toBe(
      `http://localhost:8080/api/foods/${sampleFood.id}`,
    )
    expect((fetchFn.mock.calls[0]?.[1] as RequestInit).method).toBe("PUT")
    expect(String(fetchFn.mock.calls[1]?.[0])).toBe(
      `http://localhost:8080/api/foods/${sampleFood.id}/archive`,
    )
  })

  it("upserts and clears exposures", async () => {
    const exposure = {
      foodId: sampleFood.id,
      variantKey: "bagelsaurus",
      familiarity: "safe" as const,
      source: "manual" as const,
      attemptCount: null,
      lastTriedOn: null,
      lastLiked: null,
    }
    const fetchFn = vi
      .fn()
      .mockResolvedValueOnce(
        new Response(JSON.stringify(exposure), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
      )
      .mockResolvedValueOnce(new Response(null, { status: 204 }))

    const client = new FoodsClient("http://localhost:8080", fetchFn, memoryStore())
    const upserted = await client.upsertExposure(sampleFood.id, {
      variantKey: "Bagelsaurus",
      familiarity: "safe",
    })
    await client.clearExposure(sampleFood.id, "Bagelsaurus")

    expect(upserted.variantKey).toBe("bagelsaurus")
    expect(String(fetchFn.mock.calls[0]?.[0])).toBe(
      `http://localhost:8080/api/foods/${sampleFood.id}/exposures`,
    )
    expect((fetchFn.mock.calls[0]?.[1] as RequestInit).method).toBe("PUT")
    expect(String(fetchFn.mock.calls[1]?.[0])).toBe(
      `http://localhost:8080/api/foods/${sampleFood.id}/exposures?variantKey=Bagelsaurus`,
    )
    expect((fetchFn.mock.calls[1]?.[1] as RequestInit).method).toBe("DELETE")
  })

  it("bootstraps safes via POST /api/foods/bootstrap-safes", async () => {
    const exposures = [
      {
        foodId: "aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaa04",
        variantKey: "honeycrisp",
        familiarity: "safe",
        source: "signup",
      },
    ]
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(JSON.stringify(exposures), {
        status: 200,
        headers: { "Content-Type": "application/json" },
      }),
    )
    const client = new FoodsClient("http://localhost:8080", fetchFn, memoryStore())

    const result = await client.bootstrapSafes({
      items: [
        { name: "Apples", variantKey: "Honeycrisp", sessionEligible: true },
        { name: "Goldfish", sessionEligible: false },
      ],
    })

    expect(result).toEqual(exposures)
    expect(String(fetchFn.mock.calls[0]?.[0])).toBe(
      "http://localhost:8080/api/foods/bootstrap-safes",
    )
    expect((fetchFn.mock.calls[0]?.[1] as RequestInit).method).toBe("POST")
    expect((fetchFn.mock.calls[0]?.[1] as RequestInit).body).toBe(
      JSON.stringify({
        items: [
          { name: "Apples", variantKey: "Honeycrisp", sessionEligible: true },
          { name: "Goldfish", sessionEligible: false },
        ],
      }),
    )
  })

  it("surfaces bootstrap validation errors", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(
        JSON.stringify({
          message: "At most 10 safe foods can be bootstrapped",
        }),
        {
          status: 400,
          headers: { "Content-Type": "application/json" },
        },
      ),
    )
    const client = new FoodsClient("http://localhost:8080", fetchFn, memoryStore())

    await expect(
      client.bootstrapSafes({
        items: Array.from({ length: 11 }, (_, i) => ({ name: `Food ${i}` })),
      }),
    ).rejects.toThrow("At most 10 safe foods can be bootstrapped")
  })

  it("lists adds and removes stretch targets", async () => {
    const target = {
      id: "dddddddd-dddd-dddd-dddd-dddddddddddd",
      foodId: "aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaa24",
      foodName: "Broccoli",
      variantKey: "steamed",
      createdAt: "2026-08-03T18:00:00Z",
    }
    const fetchFn = vi
      .fn()
      .mockResolvedValueOnce(
        new Response(JSON.stringify([target]), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
      )
      .mockResolvedValueOnce(
        new Response(JSON.stringify(target), {
          status: 201,
          headers: { "Content-Type": "application/json" },
        }),
      )
      .mockResolvedValueOnce(new Response(null, { status: 204 }))

    const client = new FoodsClient("http://localhost:8080", fetchFn, memoryStore())
    const listed = await client.listStretchTargets()
    const created = await client.addStretchTarget({
      foodId: target.foodId,
      variantKey: "Steamed",
    })
    await client.removeStretchTarget(target.foodId, "Steamed")

    expect(listed).toEqual([target])
    expect(created.foodName).toBe("Broccoli")
    expect(String(fetchFn.mock.calls[0]?.[0])).toBe(
      "http://localhost:8080/api/foods/stretch-targets",
    )
    expect(String(fetchFn.mock.calls[1]?.[0])).toBe(
      "http://localhost:8080/api/foods/stretch-targets",
    )
    expect((fetchFn.mock.calls[1]?.[1] as RequestInit).method).toBe("POST")
    expect(String(fetchFn.mock.calls[2]?.[0])).toBe(
      `http://localhost:8080/api/foods/stretch-targets/${target.foodId}?variantKey=Steamed`,
    )
    expect((fetchFn.mock.calls[2]?.[1] as RequestInit).method).toBe("DELETE")
  })

  it("surfaces stretch target cap errors", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(
        JSON.stringify({ message: "At most 5 stretch targets can be active" }),
        {
          status: 400,
          headers: { "Content-Type": "application/json" },
        },
      ),
    )
    const client = new FoodsClient("http://localhost:8080", fetchFn, memoryStore())

    await expect(
      client.addStretchTarget({ name: "One too many" }),
    ).rejects.toThrow("At most 5 stretch targets can be active")
  })

  it("surfaces invalid icon errors", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(JSON.stringify({ message: "Invalid icon key" }), {
        status: 400,
        headers: { "Content-Type": "application/json" },
      }),
    )
    const client = new FoodsClient("http://localhost:8080", fetchFn, memoryStore())

    await expect(
      client.create({ name: "Bad", iconKey: "nope" }),
    ).rejects.toThrow("Invalid icon key")
  })

  it("surfaces duplicate name conflicts", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(
        JSON.stringify({ message: "A food with that name already exists" }),
        {
          status: 409,
          headers: { "Content-Type": "application/json" },
        },
      ),
    )
    const client = new FoodsClient("http://localhost:8080", fetchFn, memoryStore())

    await expect(
      client.create({ name: "Watermelon", iconKey: "custom_watermelon" }),
    ).rejects.toThrow("A food with that name already exists")
  })

  it("requires a signed-in token", async () => {
    const client = new FoodsClient(
      "http://localhost:8080",
      vi.fn(),
      memoryStore(null),
    )
    await expect(client.list()).rejects.toThrow("Not signed in")
  })

  it("parses optional iconUrl when present or absent", async () => {
    const withUrl = {
      ...sampleFood,
      iconUrl: "https://cdn.example.com/foods/custom_cucumber.png",
    }
    const withoutUrl = { ...sampleFood, id: "dddddddd-dddd-dddd-dddd-dddddddddddd" }
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(JSON.stringify([withUrl, withoutUrl]), {
        status: 200,
        headers: { "Content-Type": "application/json" },
      }),
    )

    const client = new FoodsClient("http://localhost:8080", fetchFn, memoryStore())
    const foods = await client.list()

    expect(foods[0]?.iconUrl).toBe(
      "https://cdn.example.com/foods/custom_cucumber.png",
    )
    expect(foods[1]?.iconUrl).toBeUndefined()
  })
})
